File: C++/finance_calculators.cpp
#include <iostream>
#include <string> 
#include <cctype>
#include <algorithm>    
#include <cmath>

using std::cout;
using std::cin;

void calc_investment();
void calc_bond();

int main() {
    cout << "WELCOME\n-------\n";
    
    int type;
    cout << "\nPlease select an investment type: ";
    cout << "\n---------------------------------";
    cout << "\n1. INVESTMENT - to calculate the amount of interest you'll earn on your investment.";
    cout << "\n2. BOND        - to calculate the amount you'll have to pay on a home loan";
    cout << "\nEnter the option number: ";
    if (!(cin >> type)){
        cout << "\nInvalid Input. Exiting Program...";
            return 1;
    }
    
    switch(type){
    case 1: 
        calc_investment();
        break;
    case 2:
        calc_bond();
        break;
    default: 
        cout << "Invalid Option. Exiting Program..."; 
    }


    return 0;
}

void calc_investment() {
    cout << "\nInvestment Calculator";
    cout << "\n---------------------\n\n";

    float initAmount, interest;
    int years;

    cout << "How much money are you depositing? £";
    if (!(cin >> initAmount)) {
        cout << "Invalid Input detected! Exiting program...";
        return;
    }
    cout << "Enter the interest rate as a %: ";
    if (!(cin >> interest)) {
        cout << "Invalid Input detected! Exiting program...";
        return;
    }
    cout << "Enter the total length of investment (in years): ";
    if (!(cin >> years)) {
        cout << "Invalid Input detected! Exiting program...";
        return;
    }

    float finalAmount;
    std::string type;
    while (true) {
        cout << "Do you want 'simple' or 'compound' interest? ";
        cin >> type;
        std::transform(type.begin(), type.end(), type.begin(), [](unsigned char c){return std::tolower(c);});
        if (type == "simple") {
            finalAmount = initAmount * (1 + (interest / 100) * years);
            break;
        }
        else if (type == "compound") {
            finalAmount = initAmount * pow(1 + (interest / 100), years);
            break;
        }
        else {
            cout << "Invalid input! Please type either 'simple' or 'compound'.\n\n";
        }
    }

    cout << "\nInitial investment:   " << initAmount << '\n';
    cout << "Length of investment: " << years << '\n';
    cout << "Projected earnings:   " << finalAmount;    
}

void calc_bond() {
    cout << "\nBond Calculator";
    cout << "\n---------------";

    float curVal, interest;
    int months;

    cout << "\nEnter the present value: £";
    if (!(cin >> curVal)) {
        cout << "Invalid Input detected! Exiting program...";
        return;
    }
    cout << "Enter the interest rate as a %: ";
    if (!(cin >> interest)) {
        cout << "Invalid Input detected! Exiting program...";
        return;
    }
    else {
        interest = (interest / 100) / 12;
    }
    cout << "Enter the number of months you wish to make payments for: ";
    if (!(cin >> months)) {
        cout << "Invalid Input detected! Exiting program...";
        return;
    }

    float repayment = (interest * curVal) / (1- pow((1 + interest), -months));
    cout << "\nMonthly repayments: £" << repayment;

}

File: Python/finance_calculators.py
import math 

# Calculate the total amount of interestr gained on an initial investment with either
# simple or compound interest.
def calc_investment():
    print("\nInvestment Calculator")
    print("---------------------")
    try:
        init_amount = float(input("\nHow much money are you depositing: £"))
        interest = float(input("Enter the interest rate as a %: "))
        years = int(input("Enter the total length of investment (in years): "))
    except:
        print("Invlid Input detected! Exiting program.")
        return 
    
    validInp = False 
    while not validInp:
        type = input("\nDo you want 'simple' or 'compound' interest? ")
        if type.lower() == "simple" or type.lower() == "compound":
            validInp = True
        else:
            print("Invalid input! Please type either 'simple' or 'compound'.") 
    
    # Calculate results
    if type.lower() == "simple":
        final_amount = init_amount * (1 + (interest/100) * years)
    else:
        final_amount = init_amount * math.pow((1 + (interest/100)), years) 
    
    # Print results
    print("\nWith an initial investment of £{0:.2f} over {1:d} years, you will have earned £{2:.2f} in interest".format(init_amount, years, (final_amount-init_amount)))
    return 

# Calculate the total monthly repayments on a house 
def calc_bond():
    print("\nBond Calculator")
    print("---------------")
    try:
        cur_value = float(input("\nEnter the present value: £"))
        interest = float(input("Enter the % annual interest: "))
        interest = (interest / 100) / 12
        months = int(input("Enter the number of months you wish to make payments for: "))
    except:
        print("Invlid Input detected! Exiting program.")
        return 

    # Calculate and print result
    repayment = (interest * cur_value) / (1 - (1 + interest)**(-months))
    print("\nYour monthly repayments will be: £{0:.2f}".format(repayment))
    return




# Get the user to pick the type of investment they wish to calculate
print("WELCOME")
print("-------")
print("\nPlease select an investment type:")
print(  "---------------------------------")
print("1. INVESTEMENT - to calculate the amount of interest you'll earn on your investment")
print("2. BOND        - to calculate the amount you'll have to pay on a home loan")
# print("3. EXIT")
opt = int(input("Enter the integer option "))

if opt == 1:
    calc_investment()
elif opt == 2:
    calc_bond()
else:
    print("Invalid Input!")